### ISSUE TITLE
Defining a calculator community site repository format standard

### ISSUE THREAD

**Issue by TC01** [TC01/calcpkg#21]
Per conversations with Adriweb and Eeems on IRC, it was suggested that we try to write out a better standard for a repository format for calculator community sites.

Whatever is defined needs to be compatible, at the bare minimum, with the ticalc.org repository format, because that is our primary usage case. Ideally, it will be a superset of the current index file approach.

This issue will track work on such a thing and then implementation, hopefully.

It was suggested we create a Google Doc; maybe I will do that eventually but for now please leave your thoughts on [this wiki page](https://github.com/TC01/calcpkg/wiki/A-shared-repository-format-across-calculator-community-websites). EDIT: or see @adriweb's post below.

**Comment by Eeems** [comment]
Blocking #12

**Comment by adriweb** [comment]
GDocs with the archive properties : https://docs.google.com/document/d/1ZIHelCEM123lJSETyuB-Pda9W0WRBtCuGdq0w01g5Sg/edit?usp=sharing

**Comment by Eeems** [comment]
I think the scope of this should also include documenting the preferred API for using calcpkg to manage a~~n~~ ~~archive~~ repo as per #12

**Comment by adriweb** [comment]
Yeah, we should probably establish some "base API" things, like :
- Typical possible requests
- What a response can look like (different format, like JSON/XML etc.)
- Error codes/messages

The actual implementation and how the API can be queried doesn't have to be the same for everyone though, as this can certainly make much more work for the websites' maintainer. However they should still implement all the queries defined above in some way or another, and be able to reply with the info agreed upon...

I guess we can use the same document, it will be better to have something centralized.

**Comment by adriweb** [comment]
I've updated the document with a whole bunch of proposed things :)

**Comment by TC01** [comment]
I've been waiting on replying further to this because I'm not really sure how I feel about it... I'm still not really sure. But I will first say that any API that calcpkg could support should meet the following criteria:
1. It must be possible to fetch a local index of all files in the repository, and store enough basic search metadata (categories, names, etc.) to do further queries to fetch more detailed info. (The rest of the proposed attributes, I guess. They do look good, I should add). I suppose I'm thinking of "platforms" and "targets" as meta-categories, because...
2. It should be possible to maintain a tree hierarchy of the files in a repository. This is _primarily_ to maintain compatibility with ticalc.org (which I feel, regrettably or otherwise, is important because ticalc.org has an order of magnitude more files in its archives than anything else we'd potentially be supporting), and also because I don't feel like rewriting all the code that assumes this is how things work. :)

I am definitely interested in working with you guys on designing a layout/API that applications like calcpkg can use (and I really do appreciate the interest in the project!), but since my design goals are primarily driven by "how can I make the other sites like ticalc.org" instead of the other way around, if you left it all up to me I would just have you implement the ticalc.org archives and "API" and call it a day.

And because said archives are, well, ancient, you probably don't want to do that. :)

But I _would_ like to design something that's easy to write a compatibility layer for.

Just thought I'd explicitly state where I'm coming from when thinking about this.

**Comment by adriweb** [comment]
It's better to support a unified thing across multiple sites (which is what the GDocs aims to do) than currently a prehistoric thing that is an index file to parse... I mean... that's not an API!
So, I guess either ticalc implements whatever will be "standardized" in the google docs, or ticalc would have to have special support in calcpkg. ?
(By that, I mean : if (ticalc) { do some very specific things for ticalc } else { do some nice unified things for several websites at once } )

**Comment by TC01** [comment]
The odds of ticalc.org implementing whatever standard we come up with here are extremely low, I would have to say. And I would rather not turn calcpkg into "implementation of {standard}, and also we support ticalc.org", because the primary point of calcpkg was and is to support ticalc.org.

I mean, the code has already been written such that multiple repositories can behave very differently; that's not a problem. If minimal compatibility with the two points I listed happens, then a "StandardRepository" class or something can be created (I don't know what I'd call it; I guess whatever the standard gets named?) that can be instantiated multiple times by, e.g. an Omnimaga plugin and a TI-Planet plugin and possibly by others in the future. Great! This sounds like a good idea to me.

But I'm not arrogant enough to assume that a standard made here will ever propagate through all community sites (because I'm well aware of [the problems with standards](http://xkcd.com/927/)), so I wouldn't want to do a huge rewrite of the software to make my standard the default*. Even if I wanted to, I wouldn't have the time.

Some rewriting would be necessary anyway, obviously. There are bits of the software that _should_ even be rewritten. The local index files should probably not be stored as plaintext, but as bits of a database or pickled or something. But I like, generally, the distro-package-manager-model of "you don't have to go on the internet to do minimal queries about packages"...

Plus, being able to keep local indices makes things like mirroring potentially possible in the long-term future.
- what I mean here is, the rest of the software _does_ assume it can index through files in some kind of tree-like structure. That could be changed, but I'm not clear there is a huge benefit to doing.

**Comment by adriweb** [comment]
> The odds of ticalc.org implementing whatever standard we come up with here are extremely low, I would have to say

Well, I'm definitely _not_ implementing any kind of subpar "file index" thing into TI-Planet when there is[/will be] a nice API "standard" that does 9001 times what the index way can provide, just so that an unmaintained website architecture (ticalc) can serve as the base of a client... I mean, at this point, it's just a huge regression.

> the primary point of calcpkg was and is to support ticalc.org.

It could still support it thanks to the current code, but then if at least 2 websites implement the standard, then ticalc would become a minority and would have to be treated like a special case

> assume that a standard made here will ever propagate through all community sites

Well, there are basically 4... so it's definitely "possible".

> (because I'm well aware of the problems with standards)

Eh. Except that this doesn't apply since there is no standard already (hence the work being done to make one)

> I wouldn't have the time.

Well, ok, that's one argument :P But I don't believe it would take a large amount of time, though. There are only a few queries/responses being talked about so far, and a clear "Archive class" with several properties. It's really not that big.

>  being able to keep local indices makes things like mirroring potentially possible in the long-term future.

Uh ?

**Comment by TC01** [comment]
To be clear: I'm not saying you need to implement something server side that produces a master.index. I'm saying we should have API calls that, if used in the right way and in the right combination _client_ side, can be used to produce a local cache of all filenames / IDs / categories / platforms / targets on a site. The rest of the compatibility can probably be done inside calcpkg, with minimal modifications. Apologies if you misunderstood me.

(something like: "get all categories", "get all files in $category", "get all platforms", etc.?)

From my perspective, it would be the most _convenient_ if every site shipped a master.index I could just download ala the ticalc.org code, but I don't really expect you to do that, and for good reason, as you say. :)

It would be nice if categories, platforms, targets, etc. were put together in such a way that makes it not impossible to unfold the repository into a tree, but that's not a huge problem and can be dealt with.

And clearly the API would be very useful in other places-- in implementing the info command, in getting a download link, etc. All those methods are (I think?) already implemented by the Repository class so it'd be really easy to tell them to use the API instead of doing whatever hardcoded web-scraping is implemented for ticalc.org, Omnimaga, and Cemetech.

And sure, perhaps for release N+2 (assuming all of the above is for release N+1) search code is moved into the repository class and the option is provided to search using API calls directly rather than the local copies.

**Comment by adriweb** [comment]
> I'm saying we should have API calls that, if used in the right way and in the right combination client side, can be used to produce a local cache of all filenames / IDs / categories / platforms / targets on a site.

Oh, then it's all good for TI-Planet as this kind of query is [already implemented](https://github.com/TI-Planet/API#request-types) and working :)
It's also the first query proposal on the google docs ;)

And yeah, something that goes with the standard for a version n+2 isn't a problem ; I mean, first, the sites will have to implement it :P

**Comment by TC01** [comment]
> It's also the first query proposal on the google docs ;)

Oops. That's what I get for skimming the opening. :P

Then yeah, that sounds like a plan. As long as the standard has those API calls, it looks good to me. :) 

One ticalc attribute that isn't currently proposed to be supported is reviews for files; this doesn't appear to be something TI-Planet has at the moment based on a quick check? I think Omnimaga did, and Cemetech does. I am not quite sure how reviews should be handled though, if it's supported / if there are any.

**Comment by adriweb** [comment]
TI-Planet _currently_ does not have archive reviews indeed (though we have ratings, but reviews are also planned somewhat)
But that's not a problem for the API. For example, it could be an optional field. Or even a required one, and sites that don't implement that feature would respond that there is no review for this archive, or something of the sort.
Feel free to edit the doc, it's there for that reason :D
